File: phase02-todo-app/frontend/types/task.ts
export interface Task {
  id: number;
  title: string;
  description?: string;
  completed: boolean;
  priority: 'urgent' | 'high' | 'medium' | 'low';
  due_date?: string;
  tags?: string;
  user_id: string;
  created_at: string;
  updated_at: string;
  completed_at?: string;
}

export interface TaskCreate {
  title: string;
  description?: string;
  completed?: boolean;
  priority?: 'urgent' | 'high' | 'medium' | 'low';
  due_date?: string;
  tags?: string;
}

export interface TaskUpdate {
  title?: string;
  description?: string;
  completed?: boolean;
  priority?: 'urgent' | 'high' | 'medium' | 'low';
  due_date?: string;
  tags?: string;
}

export interface TaskToggleComplete {
  completed: boolean;
}

export interface TaskFilters {
  status?: 'all' | 'completed' | 'pending';
  priority?: 'low' | 'medium' | 'high';
  search?: string;
  dueDateFrom?: string;
  dueDateTo?: string;
  tag?: string;
  sortBy?: 'created_at' | 'due_date' | 'priority';
  sortOrder?: 'asc' | 'desc';
  offset?: number;
  limit?: number;
}

File: phase02-todo-app/backend/models/task.py
from sqlmodel import SQLModel, Field
from typing import Optional, List
from datetime import datetime
from sqlalchemy import Index


class TaskBase(SQLModel):
    title: str
    description: Optional[str] = None
    completed: bool = False
    priority: Optional[str] = "medium"  # Default to medium priority
    due_date: Optional[datetime] = None
    tags: Optional[str] = None  # Store tags as comma-separated string


class Task(TaskBase, table=True):
    id: int = Field(default=None, primary_key=True, sa_column_kwargs={"autoincrement": True})  # ← FIXED: autoincrement + int
    user_id: str = Field(index=True)   # user.id is STRING (uuid)
    created_at: datetime = Field(default_factory=datetime.utcnow, index=True)
    updated_at: datetime = Field(default_factory=datetime.utcnow, index=True)
    completed_at: Optional[datetime] = Field(default=None, index=True)

    __table_args__ = (Index('idx_user_completed', 'user_id', 'completed'),)


class TaskCreate(TaskBase):
    pass


class TaskUpdate(SQLModel):
    title: Optional[str] = None
    description: Optional[str] = None
    completed: Optional[bool] = None
    priority: Optional[str] = None
    due_date: Optional[datetime] = None
    tags: Optional[str] = None


class TaskPublic(TaskBase):
    id: int
    user_id: str
    created_at: datetime
    updated_at: datetime
    completed_at: Optional[datetime] = None


class TaskResponse(TaskBase):
    id: int
    user_id: str
    created_at: datetime
    updated_at: datetime
    completed_at: Optional[datetime] = None